File: sdk/java/storage/keeper_secrets_manager_storage_azure_kms/src/test/java/com/keepersecurity/secretsManager/storage/azure/AzureSessionConfigTest.java
package com.keepersecurity.secretsManager.storage.azure;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for AzureSessionConfig class
 */
class AzureSessionConfigTest {

    @Test
    void testDefaultConstructor() {
        // When
        AzureSessionConfig config = new AzureSessionConfig();

        // Then
        assertNull(config.getTenantId(), "TenantId should be null for default constructor");
        assertNull(config.getClientId(), "ClientId should be null for default constructor");
        assertNull(config.getClientSecret(), "ClientSecret should be null for default constructor");
        assertNull(config.getKeyVaultUrl(), "KeyVaultUrl should be null for default constructor");
    }

    @Test
    void testParameterizedConstructor() {
        // Given
        String tenantId = "12345678-1234-1234-1234-123456789012";
        String clientId = "87654321-4321-4321-4321-210987654321";
        String clientSecret = "test-client-secret-value";
        String keyVaultUrl = "https://test-keyvault.vault.azure.net/";

        // When
        AzureSessionConfig config = new AzureSessionConfig(
            tenantId, clientId, clientSecret, keyVaultUrl
        );

        // Then
        assertEquals(tenantId, config.getTenantId(), "TenantId should match constructor parameter");
        assertEquals(clientId, config.getClientId(), "ClientId should match constructor parameter");
        assertEquals(clientSecret, config.getClientSecret(), "ClientSecret should match constructor parameter");
        assertEquals(keyVaultUrl, config.getKeyVaultUrl(), "KeyVaultUrl should match constructor parameter");
    }

    @Test
    void testSettersAndGetters() {
        // Given
        AzureSessionConfig config = new AzureSessionConfig(
            "initial-tenant", "initial-client", "initial-secret", "https://initial.vault.azure.net/"
        );

        String newTenantId = "98765432-8765-8765-8765-987654321098";
        String newClientId = "11223344-1122-1122-1122-112233445566";
        String newClientSecret = "new-secret-value";
        String newKeyVaultUrl = "https://production-keyvault.vault.azure.net/";

        // When
        config.setTenantId(newTenantId);
        config.setClientId(newClientId);
        config.setClientSecret(newClientSecret);
        config.setKeyVaultUrl(newKeyVaultUrl);

        // Then
        assertEquals(newTenantId, config.getTenantId(), "TenantId getter should return value set by setter");
        assertEquals(newClientId, config.getClientId(), "ClientId getter should return value set by setter");
        assertEquals(newClientSecret, config.getClientSecret(), "ClientSecret getter should return value set by setter");
        assertEquals(newKeyVaultUrl, config.getKeyVaultUrl(), "KeyVaultUrl getter should return value set by setter");
    }

    @Test
    void testSettersWithNullValues() {
        // Given
        AzureSessionConfig config = new AzureSessionConfig(
            "initial-tenant", "initial-client", "initial-secret", "https://initial.vault.azure.net/"
        );

        // When - set all to null
        config.setTenantId(null);
        config.setClientId(null);
        config.setClientSecret(null);
        config.setKeyVaultUrl(null);

        // Then
        assertNull(config.getTenantId(), "TenantId should be null after setting to null");
        assertNull(config.getClientId(), "ClientId should be null after setting to null");
        assertNull(config.getClientSecret(), "ClientSecret should be null after setting to null");
        assertNull(config.getKeyVaultUrl(), "KeyVaultUrl should be null after setting to null");
    }

    @Test
    void testSettersWithEmptyStrings() {
        // Given
        AzureSessionConfig config = new AzureSessionConfig(
            "initial-tenant", "initial-client", "initial-secret", "https://initial.vault.azure.net/"
        );

        // When
        config.setTenantId("");
        config.setClientId("");
        config.setClientSecret("");
        config.setKeyVaultUrl("");

        // Then
        assertEquals("", config.getTenantId(), "TenantId should handle empty string");
        assertEquals("", config.getClientId(), "ClientId should handle empty string");
        assertEquals("", config.getClientSecret(), "ClientSecret should handle empty string");
        assertEquals("", config.getKeyVaultUrl(), "KeyVaultUrl should handle empty string");
    }

    @Test
    void testConstructorWithNullValues() {
        // When/Then - constructor should accept null values
        assertDoesNotThrow(() -> {
            AzureSessionConfig config = new AzureSessionConfig(null, null, null, null);
            assertNull(config.getTenantId());
            assertNull(config.getClientId());
            assertNull(config.getClientSecret());
            assertNull(config.getKeyVaultUrl());
        }, "Constructor should accept null values without throwing exception");
    }

    @Test
    void testKeyVaultUrlFormats() {
        // Test various Azure Key Vault URL formats
        String[] keyVaultUrls = {
            "https://test-keyvault.vault.azure.net/",
            "https://production-keyvault.vault.azure.net/",
            "https://keyvault-dev.vault.azure.net/",
            "https://my-kv.vault.azure.net",
            "https://keyvault.vault.usgovcloudapi.net/",  // Azure Government
            "https://keyvault.vault.azure.cn/"  // Azure China
        };

        for (String keyVaultUrl : keyVaultUrls) {
            // Given/When
            AzureSessionConfig config = new AzureSessionConfig(
                "12345678-1234-1234-1234-123456789012",
                "87654321-4321-4321-4321-210987654321",
                "test-secret",
                keyVaultUrl
            );

            // Then
            assertEquals(keyVaultUrl, config.getKeyVaultUrl(),
                "Should accept Key Vault URL format: " + keyVaultUrl);
        }
    }

    @Test
    void testTenantIdFormats() {
        // Test various Azure tenant ID formats
        String[] tenantIds = {
            "12345678-1234-1234-1234-123456789012",  // Standard GUID
            "common",  // Multi-tenant
            "organizations",  // Work/school accounts
            "consumers"  // Personal Microsoft accounts
        };

        for (String tenantId : tenantIds) {
            // Given/When
            AzureSessionConfig config = new AzureSessionConfig(
                tenantId,
                "87654321-4321-4321-4321-210987654321",
                "test-secret",
                "https://test-keyvault.vault.azure.net/"
            );

            // Then
            assertEquals(tenantId, config.getTenantId(),
                "Should accept tenant ID format: " + tenantId);
        }
    }
}

File: sdk/javascript/packages/azure/test/AzureSessionConfig.test.ts
import { AzureSessionConfig } from '../src/AzureSessionConfig';

describe('AzureSessionConfig', () => {
    describe('constructor', () => {
        it('should create config with valid parameters', () => {
            // Given
            const tenantId = 'test-tenant-123';
            const clientId = 'test-client-456';
            const clientSecret = 'test-secret-789';

            // When
            const config = new AzureSessionConfig(tenantId, clientId, clientSecret);

            // Then
            expect(config.tenantId).toBe(tenantId);
            expect(config.clientId).toBe(clientId);
            expect(config.clientSecret).toBe(clientSecret);
        });

        it('should handle empty string values', () => {
            // When
            const config = new AzureSessionConfig('', '', '');

            // Then
            expect(config.tenantId).toBe('');
            expect(config.clientId).toBe('');
            expect(config.clientSecret).toBe('');
        });

        it('should store all provided values correctly', () => {
            // Given
            const configs = [
                ['tenant1', 'client1', 'secret1'],
                ['aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee', 'client-guid', 'very-long-secret-key-here'],
                ['short', 'c', 's']
            ];

            configs.forEach(([tenant, client, secret]) => {
                // When
                const config = new AzureSessionConfig(tenant, client, secret);

                // Then
                expect(config.tenantId).toBe(tenant);
                expect(config.clientId).toBe(client);
                expect(config.clientSecret).toBe(secret);
            });
        });
    });

    describe('property assignment', () => {
        it('should allow modifying tenantId after creation', () => {
            // Given
            const config = new AzureSessionConfig('original-tenant', 'client', 'secret');
            const newTenantId = 'new-tenant-id';

            // When
            config.tenantId = newTenantId;

            // Then
            expect(config.tenantId).toBe(newTenantId);
            expect(config.clientId).toBe('client');
            expect(config.clientSecret).toBe('secret');
        });

        it('should allow modifying clientId after creation', () => {
            // Given
            const config = new AzureSessionConfig('tenant', 'original-client', 'secret');
            const newClientId = 'new-client-id';

            // When
            config.clientId = newClientId;

            // Then
            expect(config.tenantId).toBe('tenant');
            expect(config.clientId).toBe(newClientId);
            expect(config.clientSecret).toBe('secret');
        });

        it('should allow modifying clientSecret after creation', () => {
            // Given
            const config = new AzureSessionConfig('tenant', 'client', 'original-secret');
            const newClientSecret = 'new-secret';

            // When
            config.clientSecret = newClientSecret;

            // Then
            expect(config.tenantId).toBe('tenant');
            expect(config.clientId).toBe('client');
            expect(config.clientSecret).toBe(newClientSecret);
        });

        it('should allow setting properties to empty strings', () => {
            // Given
            const config = new AzureSessionConfig('tenant', 'client', 'secret');

            // When
            config.tenantId = '';
            config.clientId = '';
            config.clientSecret = '';

            // Then
            expect(config.tenantId).toBe('');
            expect(config.clientId).toBe('');
            expect(config.clientSecret).toBe('');
        });
    });

    describe('Azure-specific formats', () => {
        it('should accept valid Azure tenant ID formats', () => {
            // Given - various Azure tenant ID formats
            const tenantIds = [
                'aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee', // GUID format
                'contoso.onmicrosoft.com',               // Domain format
                'organizations',                          // Multi-tenant
                'consumers',                              // Personal Microsoft accounts
                'common'                                  // Both work and personal accounts
            ];

            tenantIds.forEach(tenantId => {
                // When
                const config = new AzureSessionConfig(tenantId, 'test-client', 'test-secret');

                // Then
                expect(config.tenantId).toBe(tenantId);
            });
        });

        it('should accept valid Azure client ID (Application ID) formats', () => {
            // Given - Azure Application (Client) IDs are GUIDs
            const clientIds = [
                'aaaaaaaa-1111-2222-3333-bbbbbbbbbbbb',
                '12345678-90ab-cdef-1234-567890abcdef',
                'ffffffff-ffff-ffff-ffff-ffffffffffff'
            ];

            clientIds.forEach(clientId => {
                // When
                const config = new AzureSessionConfig('test-tenant', clientId, 'test-secret');

                // Then
                expect(config.clientId).toBe(clientId);
            });
        });

        it('should accept various client secret formats', () => {
            // Given - Client secrets can be various formats
            const clientSecrets = [
                'simple-secret',
                'Very.Long~Secret_With-Special!Characters123',
                'aB3dE5fG7hI9jK1lM3nO5pQ7rS9tU1vW3xY5zA7bC9dE',
                '~Q8XYZ.123ABC-456def_789ghi'
            ];

            clientSecrets.forEach(clientSecret => {
                // When
                const config = new AzureSessionConfig('test-tenant', 'test-client', clientSecret);

                // Then
                expect(config.clientSecret).toBe(clientSecret);
            });
        });
    });
});
